drivers/spi/spi_silabs_eusart.c

```diff
@@ -137,21 +174,288 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		if (buffer) {
+			new_blk_cfg->dest_address = (uint32_t)buffer;
+			new_blk_cfg->dest_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+		} else {
+			/* Null buffer pointer means rx to null byte */
+			new_blk_cfg->dest_address = (uint32_t)&empty_buffer;
+			new_blk_cfg->dest_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		}
+	}
+	/* Setup max transfer according to requested transaction size.
+	 * Will top if bigger than the maximum transfer size.
+	 */
+	new_blk_cfg->block_size = MIN(requested_transaction_size,
+				      SPI_DMA_MAX_DESCRIPTOR_TRANSFER_SIZE);
+	return new_blk_cfg->block_size;
+}
+
+struct dma_block_config *spi_eusart_fill_data_desc(const struct spi_silabs_eusart_config *cfg,
+						   struct dma_block_config *desc,
+						   const struct spi_buf buffer[],
+						   int buffer_count,
+						   size_t transaction_len,
+						   bool is_tx)
+{
+	__ASSERT(transaction_len > 0, "Not supported");
+
+	size_t offset = 0U;
+	int i = 0;
+
+	while (i != buffer_count) {
+		if (!buffer[i].len) {
+			i++;
+			continue;
+		}
+		if (!desc) {
+			return NULL;
+		}
+		offset += spi_eusart_fill_desc(cfg, desc,
+					       buffer[i].buf ? (uint8_t *)buffer[i].buf + offset
+								: NULL,
+					       buffer[i].len - offset,
+					       is_tx);
+		if (offset == buffer[i].len) {
+			transaction_len -= offset;
+			offset = 0;
+			i++;
+		}
+		if (transaction_len) {
+			desc = desc->next_block;
+		}
+	}
+
+	while (transaction_len) {
+		if (!desc) {
+			return NULL;
+		}
+
+		transaction_len -= spi_eusart_fill_desc(cfg, desc, NULL, transaction_len, is_tx);
+		if (transaction_len) {
+			desc = desc->next_block;
+		}
+	}
+
+	desc->next_block = NULL;
+	return desc;
+}
+
+static void spi_eusart_reset_desc(struct dma_channel *channel)
+{
+	int i;
+
+	memset(channel->dma_descriptors, 0, sizeof(channel->dma_descriptors));
+	for (i = 0; i < ARRAY_SIZE(channel->dma_descriptors) - 1; i++) {
+		channel->dma_descriptors[i].next_block = &channel->dma_descriptors[i + 1];
+	}
+}
+
+static int spi_eusart_prepare_dma_transaction(const struct device *spi_dev,
+					     size_t padded_transaction_size, bool is_tx)
+{
+	struct spi_silabs_eusart_data *data = spi_dev->data;
+	const struct spi_silabs_eusart_config *cfg = spi_dev->config;
+	struct dma_channel *channel =
+		is_tx ? &data->dma_chan_tx : &data->dma_chan_rx;
+	const struct spi_buf *buffer = NULL;
+	size_t buffer_count = 0U;
+	struct dma_block_config *desc;
+	int ret = 0;
+
+	if (is_tx) {
+		buffer = data->ctx.current_tx;
+		buffer_count = data->ctx.tx_count;
+	} else {
+		buffer = data->ctx.current_rx;
+		buffer_count = data->ctx.rx_count;
+	}
+
+	spi_eusart_reset_desc(channel);
+	desc = spi_eusart_fill_data_desc(cfg, channel->dma_descriptors,
+					 buffer, buffer_count, padded_transaction_size, is_tx);
+	if (!desc) {
+		return -ENOMEM;
+	}
+
+	ret = spi_silabs_dma_config(spi_dev, channel, is_tx);
+	if (ret == -ENOMEM) {
+		LOG_ERR("DMA driver failed allocation.");
+		return ret;
+	}
```

Comment:
I am not sure of the usefulness of this `LOG_ERR()`. Why for this function and not for the `spi_eusart_fill_data_desc()`?

Reply:
You can drop the condition now.